subsys/bluetooth/audio/shell/bap.c

```diff
@@ -2381,69 +2402,110 @@ static uint16_t interval_to_sync_timeout(uint16_t interval)
 
 static bool scan_check_and_sync_broadcast(struct bt_data *data, void *user_data)
 {
-	const struct bt_le_scan_recv_info *info = user_data;
-	char le_addr[BT_ADDR_LE_STR_LEN];
+	struct bt_scan_recv_info *sr_info = (struct bt_scan_recv_info *)user_data;
 	struct bt_uuid_16 adv_uuid;
-	uint32_t broadcast_id;
 
-	if (data->type != BT_DATA_SVC_DATA16) {
-		return true;
-	}
+	switch (data->type) {
+	case BT_DATA_SVC_DATA16:
+		if (data->data_len < BT_UUID_SIZE_16 + BT_AUDIO_BROADCAST_ID_SIZE) {
+			return true;
+		}
 
-	if (data->data_len < BT_UUID_SIZE_16 + BT_AUDIO_BROADCAST_ID_SIZE) {
+		if (!bt_uuid_create(&adv_uuid.uuid, data->data, BT_UUID_SIZE_16)) {
+			return true;
+		}
+
+		if (bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_BROADCAST_AUDIO) != 0) {
+			return true;
+		}
+
+		sr_info->broadcast_id = sys_get_le24(data->data + BT_UUID_SIZE_16);
 		return true;
-	}
+	case BT_DATA_BROADCAST_NAME:
+		if (!IN_RANGE(data->data_len, BT_AUDIO_BROADCAST_NAME_LEN_MIN,
+			      BT_AUDIO_BROADCAST_NAME_LEN_MAX)) {
+			return true;
+		}
 
-	if (!bt_uuid_create(&adv_uuid.uuid, data->data, BT_UUID_SIZE_16)) {
+		utf8_lcpy(sr_info->broadcast_name, data->data, (data->data_len) + 1);
```

Comment:
As per the documentation, the last parameter of `utf8_lcpy` is the buffer size, not the data size. 
```suggestion
		utf8_lcpy(sr_info->broadcast_name, data->data, sizeof(sr_info->broadcast_name));
```

Reply:
Your suggestion is not accurate because `data->data` is not null-terminated, and we should copy only the `data->data_len` number of bytes. Therefore, I can use `strncpy()` + `utf8_trunc()` instead.

Reply:
Not sure I follow. `utf8_lcpy` ensures that the end result will be NULL terminated, and if `data->data-Len > sizeof(sr_info->broadcast_name)` then you overwrite invalid data.

Reply:
The problem here is that since `data->data` is not null-terminated, and normally `data->data_len < sizeof(sr_info->broadcast_name)`, invalid data greater than `data->data_len` will be copied to `sr_info->broadcast_name` and null-terminated at `sizeof(sr_info->broadcast_name) - 1` (according to the implementation of `utf8_lcpy` in `utf8.c`). So, if we want to use `utf8_lcpy`, the last field should be `data->data_len + 1`. Otherwise, we should use `strncpy` +` utf8_trunc` (like the implementation of `utf8_lcpy`), or `memcpy` + `utf8_trunc`.
Furthermore, `data->data_len` cannot be greater than `sizeof(sr_info->broadcast_name)` because of range checking in the previous lines.

Reply:
You make a good point. We should just 

```c
size_t size_to_cpy = MIN((data->data_len), sizeof(sr_info->broadcast_name) - 1);
memcpy(sr_info->broadcast_name, data->data, size_to_cpy);
sr_info->broadcast_name[size_to_cpy] = '\0';
utf8_trunc(sr_info->broadcast_name);
```

Reply:
Yes, `memcpy` is better. I just used `strncpy` to match the implementation of `utf8_lcpy`. I’ll change it.